src/CPUSolver.cpp

```diff
@@ -355,111 +262,184 @@ void CPUSolver::normalizeFluxes() {
-    sigma_s = material->getSigmaS();
     sigma_t = material->getSigmaT();
 
-    /* Initialize the source residual to zero */
-    _source_residuals[r] = 0.;
-    fsr_fission_source = 0.0;
+    /* Initialize the fission sources to zero */
+    fission_source = 0.0;
 
     /* Compute fission source for each group */
     if (material->isFissionable()) {
       for (int e=0; e < _num_groups; e++)
-        _fission_sources(r,e) = _scalar_flux(r,e) * nu_sigma_f[e];
+        fission_sources(r,e) = _scalar_flux(r,e) * nu_sigma_f[e];
 
-      fission_source = pairwise_sum<FP_PRECISION>(&_fission_sources(r,0),
+      fission_source = pairwise_sum<FP_PRECISION>(&fission_sources(r,0),
                                                   _num_groups);
-      fission_source *= inverse_k_eff;
+      fission_source /= _k_eff;
     }
 
-    else
-      fission_source = 0.0;
-
-    /* Compute total scattering source for group G */
+    /* Compute total (fission+scatter+fixed) source for group G */
     for (int G=0; G < _num_groups; G++) {
-      scatter_source = 0;
-
       for (int g=0; g < _num_groups; g++)
-        _scatter_sources(tid,g) = material->getSigmaSByGroupInline(g,G)
-                      * _scalar_flux(r,g);
-
-      scatter_source=pairwise_sum<FP_PRECISION>(&_scatter_sources(tid,0),
+        scatter_sources(tid,g) = material->getSigmaSByGroupInline(g,G)
+                                  * _scalar_flux(r,g);
+      scatter_source = pairwise_sum<FP_PRECISION>(&scatter_sources(tid,0),
                                                 _num_groups);
 
-      /* Set the fission source for FSR r in group G */
-      fsr_fission_source += fission_source * chi[G];
+      _reduced_sources(r,G) = fission_source * chi[G];
+      _reduced_sources(r,G) += scatter_source + _fixed_sources(r,G);
+      _reduced_sources(r,G) *= ONE_OVER_FOUR_PI / sigma_t[G];
+    }
+  }
+
+  delete [] fission_sources;
+  delete [] scatter_sources;
+}
+
 
-      _reduced_sources(r,G) = (fission_source * chi[G] + scatter_source) *
-                              ONE_OVER_FOUR_PI / sigma_t[G];
+/**
+ * @brief Computes the residual between source/flux iterations.
+ * @param res_type the type of residuals to compute 
+ *        (SCALAR_FLUX, FISSION_SOURCE, TOTAL_SOURCE)
+ * @return the average residual in each FSR
+ */
+double CPUSolver::computeResidual(residualType res_type) {
+
+  int norm;
+  double residual;
+  double* residuals = new double[_num_FSRs];
+  memset(residuals, 0., _num_FSRs * sizeof(double));
+
+  if (res_type == SCALAR_FLUX) {
+
+    norm = _num_FSRs * _num_groups;
+
+    for (int r=0; r < _num_FSRs; r++) {
+      for (int e=0; e < _num_groups; e++)
+        if (_old_scalar_flux(r,e) > 0.) {
+          residuals[r] += pow((_scalar_flux(r,e) - _old_scalar_flux(r,e)) / 
+                               _old_scalar_flux(r,e), 2);
+      }
     }
+  }
+
+  else if (res_type == FISSION_SOURCE) {
+
+    norm = _num_fissionable_FSRs * _num_groups;
```

Comment:
As in `TOTAL_SOURCE`, I would recommend removing `_num_groups` since the fission source is integrated over groups. This is solved in PR #146, but you might just fix it here.